torch/optim/swa.py

```diff
@@ -0,0 +1,371 @@
+        return {"opt_state": opt_state, "swa_state": swa_state,
+                "param_groups": param_groups}
+
+    def load_state_dict(self, state_dict):
+        r"""Loads the optimizer state.
+
+        Args:
+            state_dict (dict): SWA optimizer state. Should be an object returned
+                from a call to `state_dict`.
+        """
+        swa_state_dict = {"state": state_dict["swa_state"],
+                          "param_groups": state_dict["param_groups"]}
+        opt_state_dict = {"state": state_dict["opt_state"],
+                          "param_groups": state_dict["param_groups"]}
+        super(SWA, self).load_state_dict(swa_state_dict)
+        self._optimizer.load_state_dict(opt_state_dict)
+        self._opt_state = self._optimizer.state
+
+    def add_param_group(self, param_group):
+        r"""Add a param group to the :class:`Optimizer` s `param_groups`.
+
+        This can be useful when fine tuning a pre-trained network as frozen
+        layers can be made trainable and added to the :class:`Optimizer` as
+        training progresses.
+
+        Args:
+            param_group (dict): Specifies what Tensors should be optimized along
+            with group specific optimization options.
+        """
+        param_group['n_avg'] = 0
+        param_group['step_counter'] = 0
+        self._optimizer.add_param_group(param_group)
+
+    @staticmethod
+    def bn_update(loader, model, device=None):
+        r"""Updates BatchNorm running_mean, running_var buffers in the model.
+
+        It performs one pass over data in `loader` to estimate the activation
+        statistics for BatchNorm layers in the model.
+
+        Args:
+            loader (torch.utils.data.DataLoader): dataset loader to compute the
+                activation statistics on. Each data batch should be either a
+                tensor, or a list/tuple whose first element is a tensor
+                containing data.
+
+            model (torch.nn.Module): model for which we seek to update BatchNorm
+                statistics.
+
+            device (torch.device, optional): If set, data will be trasferred to
+                :attr:`device` before being passed into :attr:`model`.
+        """
+        if not _check_bn(model):
+            return
+        was_training = model.training
+        model.train()
+        momenta = {}
+        model.apply(_reset_bn)
+        model.apply(lambda module: _get_momenta(module, momenta))
+        n = 0
+        for input in loader:
+            if isinstance(input, (list, tuple)):
+                input = input[0]
+            b = input.size(0)
+
+            momentum = b / float(n + b)
+            for module in momenta.keys():
+                module.momentum = momentum
+
+            if device is not None:
+                input = input.to(device)
+
+            model(input)
+            n += b
+
+        model.apply(lambda module: _set_momenta(module, momenta))
+        model.train(was_training)
```

Comment:
The reason we are doing this `model.train()` here is that batchnorm layers are not updating the running statistics in eval mode. The idea is that if the model was in eval mode we go into train mode, update batchnorm, and then go back to eval mode. I don't see a way to avoid doing something like this unless we somehow change the batchnorm behavior. Are there any issues with going into train mode?

Reply:
A side effect is that Dropouts are also enabled after doing train() on the whole model...